og_ui/src/Controller/SubscriptionController.php

```diff
@@ -0,0 +1,163 @@
+ */
+
+namespace Drupal\og_ui\Controller;
+
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\Field\FieldStorageDefinitionInterface;
+use Drupal\Core\Url;
+use Drupal\user\Entity\User;
+use Drupal\user\EntityOwnerInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+use Drupal\og\Og;
+use Drupal\og\OgAccess;
+
+/**
+ * Controller for OG subscription routes.
+ */
+class SubscriptionController extends ControllerBase {
+
+  /**
+   * @param string $entity_type_id
+   * @param string|int $entity_id
+   *
+   * @return mixed
+   */
+  public function subscribe(Request $request, $entity_type_id, $entity_id) {
+    // @todo We don't need to re-validate the entity type and entity group here,
+    // as it's already been done in the access check?
+    $entity_storage = $this->entityTypeManager()->getStorage($entity_type_id);
+    $entity_access = $this->entityTypeManager()->getAccessControlHandler($entity_type_id);
+    $group = $entity_storage->load($entity_id);
+
+    $account = User::load($this->currentUser()->id());
+    $field_name = $request->query->get('field_name');
+
+    if (empty($field_name)) {
+      $field_name = og_get_best_group_audience_field('user', $account, $entity_type, $bundle);
+      if (empty($field_name)) {
+        throw new NotFoundHttpException();
+      }
+    }
+
+    // @todo Requires FieldableEntityInterface/ContentEntityInterface
+    $field = $group->getFieldDefinition($field_name);
+
+    if (empty($instance) || !$entity_access->fieldAccess('view', $field, $account)) {
+      // Field name given is incorrect, or user doesn't have access to the field.
+      throw new NotFoundHttpException();
+    }
+
+    if ($account->isAnonymous()) {
+      // Anonymous user can't request membership.
+      $destination = $this->getDestinationArray();
+
+      $user_login_url = Url::fromRoute('user.login', [], $destination);
+
+      // @todo I think this is correct? Other options are visitors or require
+      // approval both of which apply to the else instead of here.
+      if ($this->config('user.settings')->get('register') === USER_REGISTER_ADMINISTRATORS_ONLY) {
+        drupal_set_message($this->t('In order to join any group, you must <a href=":login">login</a>. After you have successfully done so, you will need to request membership again.', [':login' => $user_login_url]));
+      }
+      else {
+        $user_register_url = Url::fromRoute('user.register', [], $destination);
+        drupal_set_message($this->t('In order to join any group, you must <a href=":login">login</a> or <a href=":register">register</a> a new account. After you have successfully done so, you will need to request membership again.', [':register' => $user_register_url, ':login' => $user_login_url]));
```

Comment:
you still need to convert it to a string, just using a URL like that doesn't work :(